Refactor match state by applying the factory pattern to reduce complexity

Garanas opened the pull request:
## Description of the proposed changes

With the introduction of decapitation in https://github.com/FAForever/fa/issues/6667 the current setup for determining the victory condition gets a bit [complected](https://youtu.be/SxdOUGdseq4?si=7sMI1igw3CGiuDWo&t=2544). By applying a factory pattern we can reduce the state to a minimum and create more meaning and order.

## Testing done on the proposed changes


## Additional context



## Checklist

- [x] Changes are annotated, including comments where useful
- [x] Changes are documented in a changelog snippet according to the [guidelines](https://faforever.github.io/fa/development/changelog#format-of-a-snippet).
- [ ] Request 2-3 reviewers from the [list of reviewers and their areas of knowledge](https://github.com/FAForever/fa/blob/deploy/fafdevelop/CONTRIBUTING.md#reviewers).

Garanas committed:
Add an alias for the victory condition options

Garanas committed:
Introduce an abstract class to manage victory conditions

Garanas committed:
Implement the existing conditions

Garanas committed:
Implement the factory pattern

Garanas committed:
Clean up the match state

lL1l1 reviewed (commented):
I think there should be a class that handles the logic of everything except decapitation, since only self.UnitCategories is changed in those.

Garanas committed:
Update license

Garanas committed:
Upvalue table functions for performance

Garanas committed:
Factor out the check whether a unit counts towards the victory condition

Garanas committed:
Upvalue for performance

Garanas committed:
Update function reference to `BrainHasEligibleUnits`

Garanas committed:
Introduce a separate and re-useable unit condition

Garanas commented:
@lL1l1  Applied it in https://github.com/FAForever/fa/pull/6766/commits/7a3a88d6f88d95f6e2a4332e47a9413c72e99c8e . I understand where you come from, it's a balance:

- By introducing a single abstraction for multiple conditions we run the risk of a [leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction).
- By introducing a single abstraction we run the risk that it's not trivial to figure out what conditions it matches.
- By introducing a single abstraction, we move the responsibility of _what units_ trigger the condition to the abstract factory.

In this example - it is very, very unlikely that the behavior of Assasination, Supremacy and Annihilation would ever change. And therefore I think the argument of a leaky abstraction does not apply. In practice however - these things change!

lL1l1 commented on lua/sim/Matchstate/AbstractVictoryCondition.lua:
```diff
@@ -0,0 +1,230 @@
+    GetArmyBrains = function(self)
+        local participatingArmyBrains = {}
+        for k, aiBrain in ArmyBrains do
+            if not (ArmyIsOutOfGame(aiBrain:GetArmyIndex() or aiBrain:IsDefeated())) then
```
- `ArmyIsOutOfGame` had its parentheses around the entire `or` statement
- `IsDefeated` already checks `ArmyIsOutOfGame`:
https://github.com/FAForever/fa/blob/e16b829ee2603450988f283e7ee1776026c53061/lua/aibrain.lua#L411-L417
```suggestion
            if not aiBrain:IsDefeated() then
```

lL1l1 commented on lua/sim/Matchstate/AbstractVictoryCondition.lua:
```diff
@@ -0,0 +1,230 @@
+    --- A utility function that retrieves all brains that are still participating in the match.
+    ---@param self AbstractVictoryCondition
+    ---@return AIBrain[]
+    GetArmyBrains = function(self)
```
The function name implies this gets all army brains. Maybe it should be something like `GetPlayingArmyBrains` or "Eligible" or "InPlay".

lL1l1 commented on lua/sim/Matchstate/AbstractVictoryCondition.lua:
```diff
@@ -0,0 +1,230 @@
+    --- Ends the game.
+    ---@param self AbstractVictoryCondition
+    EndGameThread = function(self)
+        WaitTicks(30)
```
Why not `WaitSeconds` to match the comment above?

Garanas replied:
Applied in https://github.com/FAForever/fa/pull/6766/commits/26a33782a23213715aab80300cf5c21be6642fdd

lL1l1 commented on lua/sim/Matchstate/AbstractVictoryCondition.lua:
```diff
@@ -0,0 +1,230 @@
+        end
+
+        Sync.GameEnded = true
+        WaitTicks(1)
```
Do we know the purpose of this wait?

Also previously it was `WaitSeconds(0.1)` which is `WaitTicks(2)`?

Garanas replied:
I have no idea. I'll leave it in for now.

lL1l1 commented on lua/sim/Matchstate/AbstractVictoryCondition.lua:
```diff
@@ -0,0 +1,230 @@
+    ---@param self AbstractVictoryCondition
+    ---@param aiBrain AIBrain
+    ToObserver = function(self, aiBrain)
+        local brainIndex = aiBrain:GetArmyIndex()
```
```suggestion
        local brainIndex = aiBrain.Army
```
(and for all other `aiBrain:GetArmyIndex()` calls too)

Garanas replied:
Applied in https://github.com/FAForever/fa/pull/6766/commits/1387401439bdd6bc0afc6e90088d3291e0bd1561

lL1l1 commented on lua/sim/Matchstate/AbstractVictoryCondition.lua:
```diff
@@ -0,0 +1,230 @@
+    --- Utility function for the logic to remove an army from the game by turning it into an observer.
+    ---@param self AbstractVictoryCondition
+    ---@param aiBrain AIBrain
+    ToObserver = function(self, aiBrain)
```
I no longer see the `SetCommandSource` part, where humans lose permission to do anything, does that matter?

Garanas replied:
Re-introduced it with https://github.com/FAForever/fa/pull/6766/commits/506ac5ec22aed2eadc784605a0f61cbc19121d99

lL1l1 commented on lua/sim/Matchstate/UnitCondition.lua:
```diff
@@ -0,0 +1,91 @@
+
+---@class DecapitationCondition : AbstractVictoryCondition
+---@field UnitCategories EntityCategory
+DecapitationCondition = Class(AbstractVictoryCondition) {
```
```suggestion
UnitCondition = Class(AbstractVictoryCondition) {
```

lL1l1 commented on lua/sim/Matchstate/UnitCondition.lua:
```diff
@@ -0,0 +1,91 @@
+---@param entityCategories EntityCategory
+---@return DecapitationCondition
+CreateUnitCondition = function(entityCategories)
+    return DecapitationCondition(entityCategories)
```
```suggestion
    return UnitCondition(entityCategories)
```

lL1l1 commented on lua/sim/Matchstate/UnitCondition.lua:
```diff
@@ -0,0 +1,91 @@
+
+---@param entityCategories EntityCategory
+---@return DecapitationCondition
+CreateUnitCondition = function(entityCategories)
```
Instead of creating these functions you could add an `---@operator call(EntityCategory): DecapitationCondition` to the class annotation to make intellisense work, albeit manually (generics don't work in our outdated version).

Garanas replied:
Great find, applied it in https://github.com/FAForever/fa/pull/6766/commits/eb05d058399686fd388b5dc2ec8029754a5e6a86

lL1l1 commented on lua/sim/Matchstate/DecapitationCondition.lua:
```diff
@@ -0,0 +1,100 @@
+        end
+
+        -- check if defeated brains still have allies with ACUs
+        local decapitatedBrains = {}
```
This table shouldn't be necessary, you can just remove brains from defeated brains when they have an ally. Using `table.getn` for iteration does make it annoying though, you have to do this trick:
```lua
local n = TableGetn(t)
for i = n, 1, -1 do
    local v = t[i]
    if cond() then
        t[i], t[n] = t[n], nil
        n = n - 1
    end
end
```
which is hard to read lol so maybe its easier to just iterate with a foreach and set keys to `nil`.

lL1l1 replied:
Also if you want to aim for peak optimization you can track the `n` of tables as you insert into them from the very start, since they're not going to be edited outside this scope.

Garanas replied:
There's no performance concerns here. There's always only one instance. Readability is more favorable in my opinion.

lL1l1 commented on lua/sim/Matchstate/DecapitationCondition.lua:
```diff
@@ -0,0 +1,100 @@
+            for l = 1, TableGetn(aliveBrains) do
+                local aliveBrain = aliveBrains[l]
+                if IsAlly(defeatedBrain:GetArmyIndex(), aliveBrain:GetArmyIndex()) then
+                    TableInsert(aliveBrains, defeatedBrain)
```
The decapitated brains table is not being updated in any way.

lL1l1 commented on lua/sim/Matchstate/DecapitationCondition.lua:
```diff
@@ -0,0 +1,100 @@
+local TableGetn = table.getn
+local TableInsert = table.insert
+
+---@class DecapitationCondition : AbstractVictoryCondition
```
Class descriptions seem like a good idea.
```suggestion
--- Victory condition where brains are considered defeated if they have no allied ACUs alive.
---@class DecapitationCondition : AbstractVictoryCondition
```

lL1l1 commented on lua/sim/Matchstate/VictoryConditionFactory.lua:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    -- default to sandbox
+    WARN("Unknown victory condition type: " .. victoryConditionType)
```
concatenation operator errors with non-strings.
```suggestion
    WARN("Unknown victory condition type: " .. tostring(victoryConditionType))
```

lL1l1 commented on lua/sim/Matchstate/AbstractVictoryCondition.lua:
```diff
@@ -0,0 +1,230 @@
+
+        -- only units that are not recalling are taken into account.
+
+        if unit.RecallingAfterDefeat then
```
I think this should be refactored to something like `IsNotEligibleForVictory` in the future.

lL1l1 reviewed (changes requested):
> By introducing a single abstraction we run the risk that it's not trivial to figure out what conditions it matches.

I think the 3 files with only units being different makes it more difficult than looking at what the Factory is doing.

> By introducing a single abstraction, we move the responsibility of what units trigger the condition to the abstract factory.

I'm not sure how this matters, whatever creates a victory instance can modify the units through class fields anyway since we don't implement private fields in lua (not that its possible with the debug library being accessible).

Speaking of "whatever creates a victory instance", I don't think the factory is necessary because it would only be called once due to its game-wide, engine-behavior effects. All of its code could just be in MatchState's `Setup` function.

Garanas committed:
Update lua/sim/Matchstate/AbstractVictoryCondition.lua

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Garanas committed:
Update lua/sim/Matchstate/DecapitationCondition.lua

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Garanas committed:
Update lua/sim/Matchstate/VictoryConditionFactory.lua

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Garanas committed:
Fix condition when unit is eligible

Garanas committed:
Remove unused parameter for `GetListOfUnits`

Garanas committed:
Add base cases when determining whether game is forfeit or won

Garanas committed:
Stop processing the game state once the game has ended

Garanas committed:
Fix a copy-paste error in the factory

Garanas committed:
Fix bugs with the decapitation game mode

Garanas committed:
Apply fixes to other victory conditions

Garanas committed:
Use `SPEW` instead of `LOG`

Garanas committed:
Remove victory conditions that are replaced by the generic unit category victory condition

Garanas committed:
Apply singleton pattern

This provides a stable interface to retrieve the one and only valid instance.

Garanas committed:
Use the singleton pattern directly

Garanas committed:
Rename folder from `MatchState` to `VictoryCondition`

Garanas committed:
Fix a copy-paste error

Garanas committed:
Fix typo in name for sandbox condition

Garanas committed:
Improve naming conventions and add comments

Garanas committed:
Remove the old match state file

Garanas committed:
Add a snippet

Garanas commented:
> Speaking of "whatever creates a victory instance", I don't think the factory is necessary because it would only be called once due to its game-wide, engine-behavior effects. All of its code could just be in MatchState's `Setup` function.

I disagree because it dissolves the pattern. And by dissolving the pattern it makes it harder to reason about.

I've reworked it a bit. I think it's more clear now, and ready to be merged in. I also checked whether coop interact with these functions and they don't.

Garanas committed:
use wait seconds instead of wait ticks to better convey meaning

Garanas committed:
Use `.Army` instead of `:GetArmyIndex`

Garanas committed:
Exclude civilian brains from victory conditions

Garanas committed:
Document the `SetCommandSource` engine function

Garanas committed:
Re-introduce the command source logic

This appears a bit magical at the moment. I have no idea what this is supposed to do. The (intended) behavior and its side effects are not documented.

Garanas committed:
Add operator annotations to return the correct type when instantiating

Garanas committed:
Remove the create functions

Garanas committed:
Rename function to find eligible brains and factor out the condition

Garanas committed:
Rename function to find eligible brains and factor out the condition
